### ISSUE TITLE
Infinite loop can occur looking up references when re-exports are used

### ISSUE THREAD

**Issue by christopherthielen** [TypeStrong/typedoc#1186]
- [x] I have checked [issues with bug label](https://github.com/TypeStrong/typedoc/labels/bug) and found no duplicates

## Expected Behavior
Typedoc should handle re-exports and circular dependencies

## Actual Behavior

Typedoc hangs while trying to resolve a re-exported reference if there is a circular dependency between directories.

## Steps to reproduce the bug

Create the test case using bash:
```
mkdir dir1;
mkdir dir2;
echo "import './dir1/file1'" > index.ts
echo "export * from './file1'" > dir1/index.ts
echo "import { CONSTANT_PKG_2 } from '../dir2'" > dir1/file1.ts
echo "export const CONSTANT_PKG_1 = 1" >> dir1/file1.ts
echo "export * from './file2'" > dir2/index.ts
echo "import { CONSTANT_PKG_1 } from '../dir1'" > dir2/file2.ts
echo "export const CONSTANT_PKG_2  = 2" >> dir2/file2.ts
```

Run typedoc 0.16

`npx typedoc@0.16.7 ./index.ts`

```
npx: installed 40 in 3.603s

Using TypeScript 3.7.5 from /Users/xxxxx/.npm/_npx/76441/lib/node_modules/typedoc/node_modules/typescript/lib
No 'out' or 'json' option has been set
The './docs' directory has be set as the output location by default
Rendering [=================-----------------------] 42%
```

If you debug this node process, you will find it's hung in an infinite loop in `tryGetTargetReflectionDeep`:

<img width="856" alt="Screen Shot 2020-01-20 at 8 07 10 PM" src="https://user-images.githubusercontent.com/2053478/72775188-470dc380-3bc2-11ea-9f5a-382ef2c6c6c7.png">

Note that this works with typedoc 0.15.x:
```
npx typedoc@0.15.8 --module commonjs ./index.ts
```

```
npx: installed 40 in 3.48s

Using TypeScript 3.7.5 from /Users/xxxxx/.npm/_npx/76767/lib/node_modules/typedoc/node_modules/typescript/lib
No 'out' or 'json' option has been set
The './docs' directory has be set as the output location by default
Rendering [========================================] 100%

Documentation generated at /Users/xxxxx/projects/tt/docs
```

## Environment
  - Typedoc version: 0.16.7
  - Node.js version: 10.15.3
  - OS: MacOS

**Comment by Gerrit0** [comment]
I could have sworn I saw that and fixed it when first figuring out the themes :/ I'll try to fix this tomorrow.

**Comment by Gerrit0** [comment]
Fixed in 0.16.8
